In [2]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup


In [109]:
#pip3 install -U googlemaps
import googlemaps
import json

gmaps = googlemaps.Client(key='AIzaSyBHPgUusU8uUrhtl6I7l7SqSPzKZi2O7NA')

## Get all major area and their pincodes of West Bengal, India from www.chennaiiq.com

In [110]:
def get_pin(page,Area_Name,Pin_Code) :
    wiki_link='http://www.chennaiiq.com/india/pincode/index.asp?id=35&state_name=West%20Bengal&page='+page
    raw_wiki_page_content = requests.get(wiki_link).text
    #print(wiki_link)
    raw_wiki_page_content_xml = BeautifulSoup(raw_wiki_page_content,'html.parser')
    table = None
    for tab_cell in raw_wiki_page_content_xml.find_all('table', class_='TBox'):
        table = tab_cell
    line_number = 1
    for tr_cell in table.find_all('tr'): 
        if line_number > 1 :
            column_number = 1
            for td_cell in tr_cell.find_all('td'):
                if column_number == 3 :
                    #print(td_cell.text)
                    Area_Name.append(td_cell.text)
                if column_number == 5 :
                    Pin_Code.append(td_cell.text)                    
                column_number = column_number + 1
        line_number = line_number + 1
    #print('line_number:'+str(line_number))
    #print('Area_Name size:'+str(len(Area_Name)))
    #print('Pin_Code size:'+str(len(Pin_Code)))
    

In [111]:
page_list = ['A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z']
Area_Name = []
Pin_Code = []
for p in page_list:
    get_pin(p,Area_Name,Pin_Code)
kolkata_df = pd.DataFrame({"AreaName":Area_Name,"PinCode":Pin_Code})
kolkata_df

AreaName PinCode
0       Ananda Bazar Patrika  700001
1               Amrita Bazar  700003
2                  Ahiritola  700005
3                    Archana  700007
4          Archarya Jagadish  700020
5     Alipore Bodyguard Line  700027
6        Alipore Civil Court  700027
7     Alipore District Board  700027
8                Alipore H O  700027
9         Alipore Store Yard  700027
10                 Alambazar  700035
11                Ashokegarh  700035
12    Abinash Choudhury Lane  700046
13                  Ariadaha  700057
14                  Agarpara  700109
15                  Agarpara  700109
16                Anandapuri  700122
17                  Abdalpur  700155
18                Andul Road  711103
19       A Road (Bamongachi)  711106
20           Abhoy Guha Road  711204
21                 Anantapur  711301
22               Andul Mauri  711302
23                      Amta  711401
24                 Adconagar  712121
25             Ayemanawabpur  712155
26                    Antpar  712424
27            Arambagh Court  712601
28              Arambagh H O  712601
29         Agricultural Firm  713101
...                      ...     ...
2042                Uluberia  711315
2043               Uttarpara  712258
2044         Uttarpara Bazar  712258
2045            Uniatatarpur  713173
2046                Ushagram  713303
2047       Upper Chellidanga  713304
2048   Ukhra Puratan Hattala  713363
2049                   Ukhra  713363
2050                 Uchalon  713427
2051                     Ura  713433
2052       Uttar Soutan Chak  721664
2053                Ukilpara  741101
2054                   Usthi  743375
2055            Uttar Raipur  743389
2056             Vishyasabha  700007
2057            Viviekanagar  700075
2058               Vip Nagar  700100
2059        Vivekananda Math  700120
2060       Vivekananda Pally  700139
2061       Vivekananda Pally  700139
2062               V C Mills  711107
2063               V K Nagar  713210
2064          Vivekandanagar  723147
2065      Vidyalaya Mohanpur  741252
2066            W B Assembly  700001
2067        Writers Building  700001
2068                Watgunge  700023
2069            W B G S Camp  700062
2070       West Congresspara  735101
2071             Zone Colony  713201

[2072 rows x 2 columns]

## There are multiple important areas having the same pincode. Merging area names on pincode

In [112]:
kolkata_df = kolkata_df.groupby('PinCode').agg({'AreaName': ', '.join}).reset_index()
kolkata_df

PinCode                                           AreaName
0     700001  Ananda Bazar Patrika, Braborne Road, Council H...
1     700002  Cossipore Gun Factory, Cossipore H O, Indian R...
2     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika
3     700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...
4     700005                                Ahiritola, Hatkhola
5     700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...
6     700007  Archana, Barabazar H O, College Squire, Kalake...
7     700008                             Barisha, Purba Barisha
8     700009                                  Sarani Rarsibagan
9     700010          Beleghata H O, Joranandir, Subhas Sarobar
10    700011                                        Narkeldanga
11    700012                             Bow Bazar, Sankaritala
12    700013                              Dharamtala, Mott Lane
13    700014  Bamboovilla, Intally, Linton Street, Sealdah, ...
14    700015              Seal Lane, Sales Tax Building, Tengra
15    700016             Elliot Road, Madrassa, Park Street H O
16    700017  Circus Avenue, Jhowtala, Park Circus, Shakespe...
17    700018                             Bartala, Rabindranagar
18    700019  Ballygunge, Ballygunge R S, Ballygunge Science...
19    700020  Archarya Jagadish, Chandra Bose Road, Gokhale ...
20    700021                                       Fort Willium
21    700022                              Bakery Road, Hastings
22    700023  Kidderpore, Manstala, Market, Raja Janakinath,...
23    700024  Garden Reach, Keshoram Cotton Mills, Paharpur ...
24    700025  Bhawanipore, Lansdown Market, Ranshet Wadi, Ra...
25    700026  Keoratala, Kalighat, Kalimandir, Protisthan, R...
26    700027  Alipore Bodyguard Line, Alipore Civil Court, A...
27    700028  Dum Dum, Jungipara Salgs Nager Bzr, Jugipara S...
28    700029  Dover Lane, Lake Market, Rash Behari Avenue, S...
29    700030              Ghughudanga, Purbe Sinthee, Sethbagan
...      ...                                                ...
1355  743440                                              Metia
1356  743441                                           Minakhan
1357  743442                                       Nazathtkhola
1358  743443                                            Hadipur
1359  743501                                 Achipur, Bansdroni
1360  743502                                            Bhangar
1361  743503                                 Bishnupur Baruipur
1362  743504                                              Falta
1363  743507                                          Jairampur
1364  743512                                         I I M Joka
1365  743513                                           Fathepur
1366  743514                                           Chaberia
1367  743517                                              Bodra
1368  743601                                           Gokarnee
1369  743602                                         Rangabelia
1370  743603                                            Kautala
1371  743604                                          Gilarchat
1372  743605                                              Nalua
1373  743606                                         Gangasagar
1374  743607                                          Jalaberia
1375  743608                                   Bakultala 24 Pgs
1376  743609                                            Hamsuri
1377  743610                                            Madarat
1378  743611                                       Pathan Khali
1379  743613                                        South Garia
1380  743701                                              Palla
1381  743702                                           Rajibpur
1382  743703                                           Tentulia
1383  743704                                               Guma
1384  743705                                          Petrapole

[1385 rows x 2 columns]

In [113]:
kolkata_df.to_csv('wb_area_to_pin.csv')

## Fetch pincode wise Latitudes and Longitudes using google geocode

In [116]:
Area_Name = []
Pin_Code = []
Lat = []
Lng = []
for index, row in kolkata_df.iterrows():
    PinCode_temp = row['PinCode']
    PinCode_temp = PinCode_temp.strip()
    AreaName_temp = row['AreaName']
    AreaName_temp = AreaName_temp.strip()
    Pin_Code.append(PinCode_temp)
    Area_Name.append(AreaName_temp)
    
    try:
        geocode_result = gmaps.geocode(PinCode_temp+', West Bengal, India')
        geocode_result_s = str(geocode_result)
        geocode_result_s = geocode_result_s.replace("'","\"")
        geocode_result_s = geocode_result_s[1:len(geocode_result_s)-1]
        #print(geocode_result_s)
        geocode_result_j = json.loads(geocode_result_s)
        lat = geocode_result_j["geometry"]["location"]["lat"]
        lng = geocode_result_j["geometry"]["location"]["lng"]
        Lat.append(lat)
        Lng.append(lng)
        print("PinCode_temp:"+PinCode_temp+" lat:"+str(lat))
    except:
        Lat.append(-1)
        Lng.append(-1) 
        print("PinCode_temp:"+PinCode_temp+" lat:"+str(-1))
    
    

PinCode_temp:700001 lat:22.5736132
PinCode_temp:700002 lat:22.6189402
PinCode_temp:700003 lat:22.6038219
PinCode_temp:700004 lat:22.5983206
PinCode_temp:700005 lat:22.5980678
PinCode_temp:700006 lat:22.5890726
PinCode_temp:700007 lat:22.5825574
PinCode_temp:700008 lat:22.4811681
PinCode_temp:700009 lat:22.5766749
PinCode_temp:700010 lat:22.5630255
PinCode_temp:700011 lat:22.577486
PinCode_temp:700012 lat:22.5683534
PinCode_temp:700013 lat:22.5653004
PinCode_temp:700014 lat:22.5596244
PinCode_temp:700015 lat:22.5550397
PinCode_temp:700016 lat:22.5535048
PinCode_temp:700017 lat:22.5409396
PinCode_temp:700018 lat:22.5360345
PinCode_temp:700019 lat:22.5279995
PinCode_temp:700020 lat:22.5377634
PinCode_temp:700021 lat:22.5636078
PinCode_temp:700022 lat:22.5482461
PinCode_temp:700023 lat:22.5317391
PinCode_temp:700024 lat:22.5484801
PinCode_temp:700025 lat:22.5301479
PinCode_temp:700026 lat:22.5187493
PinCode_temp:700027 lat:22.5259613
PinCode_temp:700028 lat:22.6319864
PinCode_temp:700029 l

PinCode_temp:712202 lat:22.7350902
PinCode_temp:712203 lat:22.745277
PinCode_temp:712204 lat:22.7631477
PinCode_temp:712205 lat:-1
PinCode_temp:712208 lat:-1
PinCode_temp:712221 lat:22.8904837
PinCode_temp:712222 lat:22.795783
PinCode_temp:712223 lat:22.7893618
PinCode_temp:712232 lat:22.6739913
PinCode_temp:712233 lat:22.6858562
PinCode_temp:712234 lat:22.6975192
PinCode_temp:712235 lat:22.7007743
PinCode_temp:712245 lat:22.6722892
PinCode_temp:712246 lat:22.6976706
PinCode_temp:712247 lat:22.657546
PinCode_temp:712248 lat:22.7230606
PinCode_temp:712258 lat:22.6665072
PinCode_temp:712301 lat:22.9064946
PinCode_temp:712302 lat:22.9541364
PinCode_temp:712303 lat:23.025826
PinCode_temp:712304 lat:22.7176705
PinCode_temp:712305 lat:22.9202915
PinCode_temp:712306 lat:22.7475076
PinCode_temp:712307 lat:22.9541364
PinCode_temp:712308 lat:23.0251223
PinCode_temp:712309 lat:-1
PinCode_temp:712365 lat:-1
PinCode_temp:712401 lat:22.8335914
PinCode_temp:712402 lat:22.9534269
PinCode_temp:712403 l

PinCode_temp:721131 lat:22.3290103
PinCode_temp:721133 lat:22.0960205
PinCode_temp:721134 lat:22.42703
PinCode_temp:721136 lat:22.4475781
PinCode_temp:721137 lat:22.3801321
PinCode_temp:721138 lat:-1
PinCode_temp:721139 lat:22.4246049
PinCode_temp:721140 lat:22.2563757
PinCode_temp:721141 lat:22.4214722
PinCode_temp:721142 lat:22.361173
PinCode_temp:721144 lat:22.1492874
PinCode_temp:721145 lat:22.2520451
PinCode_temp:721146 lat:22.5201924
PinCode_temp:721147 lat:22.6576777
PinCode_temp:721148 lat:22.5515399
PinCode_temp:721149 lat:22.3260613
PinCode_temp:721150 lat:22.5293382
PinCode_temp:721151 lat:22.4085707
PinCode_temp:721152 lat:22.3533991
PinCode_temp:721153 lat:22.5281071
PinCode_temp:721154 lat:22.4444461
PinCode_temp:721155 lat:22.2339721
PinCode_temp:721156 lat:22.4946471
PinCode_temp:721157 lat:22.822032
PinCode_temp:721158 lat:22.4921032
PinCode_temp:721159 lat:21.9979679
PinCode_temp:721160 lat:22.3760968
PinCode_temp:721161 lat:22.1806988
PinCode_temp:721162 lat:-1
PinCo

PinCode_temp:723164 lat:-1
PinCode_temp:723165 lat:-1
PinCode_temp:723166 lat:-1
PinCode_temp:723167 lat:-1
PinCode_temp:723168 lat:23.4099264
PinCode_temp:723201 lat:23.4066872
PinCode_temp:723202 lat:23.3550585
PinCode_temp:723212 lat:23.2578534
PinCode_temp:723213 lat:23.4511897
PinCode_temp:723214 lat:-1
PinCode_temp:723215 lat:23.5351601
PinCode_temp:723216 lat:-1
PinCode_temp:731101 lat:23.8985373
PinCode_temp:731102 lat:23.8509789
PinCode_temp:731103 lat:23.9405374
PinCode_temp:731121 lat:23.779944
PinCode_temp:731122 lat:-1
PinCode_temp:731123 lat:23.8122168
PinCode_temp:731124 lat:23.7184163
PinCode_temp:731125 lat:23.8094263
PinCode_temp:731126 lat:23.9555481
PinCode_temp:731127 lat:24.005052
PinCode_temp:731128 lat:23.8122168
PinCode_temp:731129 lat:23.851906
PinCode_temp:731130 lat:23.989674
PinCode_temp:731131 lat:-1
PinCode_temp:731132 lat:23.9945714
PinCode_temp:731133 lat:23.7519125
PinCode_temp:731134 lat:-1
PinCode_temp:731135 lat:-1
PinCode_temp:731136 lat:23.989674


PinCode_temp:735207 lat:26.8761833
PinCode_temp:735208 lat:26.7837705
PinCode_temp:735209 lat:26.8260272
PinCode_temp:735210 lat:26.5810759
PinCode_temp:735211 lat:26.5353118
PinCode_temp:735212 lat:26.6754069
PinCode_temp:735213 lat:26.6416086
PinCode_temp:735214 lat:26.6304055
PinCode_temp:735215 lat:26.7423812
PinCode_temp:735216 lat:26.6174984
PinCode_temp:735217 lat:26.630853
PinCode_temp:735219 lat:26.6856417
PinCode_temp:735220 lat:26.7715147
PinCode_temp:735221 lat:26.8499379
PinCode_temp:735222 lat:26.8725934
PinCode_temp:735223 lat:26.9095776
PinCode_temp:735224 lat:26.5333404
PinCode_temp:735225 lat:26.8750944
PinCode_temp:735226 lat:26.6312095
PinCode_temp:735227 lat:26.725419
PinCode_temp:735228 lat:26.7711229
PinCode_temp:735229 lat:26.896873
PinCode_temp:735230 lat:26.7563574
PinCode_temp:735231 lat:26.9560074
PinCode_temp:735232 lat:26.7375049
PinCode_temp:735233 lat:26.6865653
PinCode_temp:735234 lat:26.7781673
PinCode_temp:735235 lat:-1
PinCode_temp:735236 lat:26.7563

PinCode_temp:742307 lat:-1
PinCode_temp:742401 lat:23.7871076
PinCode_temp:742402 lat:23.7701535
PinCode_temp:742403 lat:-1
PinCode_temp:742404 lat:23.8116107
PinCode_temp:742405 lat:23.9785699
PinCode_temp:743102 lat:-1
PinCode_temp:743122 lat:22.7774416
PinCode_temp:743123 lat:22.8448034
PinCode_temp:743124 lat:23.07405
PinCode_temp:743125 lat:22.8462275
PinCode_temp:743126 lat:22.876356
PinCode_temp:743127 lat:22.8256465
PinCode_temp:743128 lat:22.8449267
PinCode_temp:743129 lat:22.8631342
PinCode_temp:743133 lat:22.8185103
PinCode_temp:743134 lat:22.94445
PinCode_temp:743135 lat:22.9123841
PinCode_temp:743136 lat:22.9251673
PinCode_temp:743144 lat:22.7955159
PinCode_temp:743145 lat:22.9376162
PinCode_temp:743155 lat:22.198294
PinCode_temp:743165 lat:22.8951195
PinCode_temp:743166 lat:22.8943145
PinCode_temp:743175 lat:-1
PinCode_temp:743176 lat:-1
PinCode_temp:743178 lat:22.697172
PinCode_temp:743187 lat:-1
PinCode_temp:743188 lat:-1
PinCode_temp:743193 lat:22.947561
PinCode_temp:7

In [117]:
kolkata_df = pd.DataFrame({"AreaName":Area_Name,"PinCode":Pin_Code,"Latitude":Lat,"Longitude":Lng})
kolkata_df

AreaName PinCode   Latitude  \
0     Ananda Bazar Patrika, Braborne Road, Council H...  700001  22.573613   
1     Cossipore Gun Factory, Cossipore H O, Indian R...  700002  22.618940   
2      Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  700003  22.603822   
3     R G Kar M C Hospital, Shyambazar, Shyambazar M...  700004  22.598321   
4                                   Ahiritola, Hatkhola  700005  22.598068   
5     Beadon Street, Bidhan Sarani, Jorasanko, Manik...  700006  22.589073   
6     Archana, Barabazar H O, College Squire, Kalake...  700007  22.582557   
7                                Barisha, Purba Barisha  700008  22.481168   
8                                     Sarani Rarsibagan  700009  22.576675   
9             Beleghata H O, Joranandir, Subhas Sarobar  700010  22.563025   
10                                          Narkeldanga  700011  22.577486   
11                               Bow Bazar, Sankaritala  700012  22.568353   
12                                Dharamtala, Mott Lane  700013  22.565300   
13    Bamboovilla, Intally, Linton Street, Sealdah, ...  700014  22.559624   
14                Seal Lane, Sales Tax Building, Tengra  700015  22.555040   
15               Elliot Road, Madrassa, Park Street H O  700016  22.553505   
16    Circus Avenue, Jhowtala, Park Circus, Shakespe...  700017  22.540940   
17                               Bartala, Rabindranagar  700018  22.536034   
18    Ballygunge, Ballygunge R S, Ballygunge Science...  700019  22.527999   
19    Archarya Jagadish, Chandra Bose Road, Gokhale ...  700020  22.537763   
20                                         Fort Willium  700021  22.563608   
21                                Bakery Road, Hastings  700022  22.548246   
22    Kidderpore, Manstala, Market, Raja Janakinath,...  700023  22.531739   
23    Garden Reach, Keshoram Cotton Mills, Paharpur ...  700024  22.548480   
24    Bhawanipore, Lansdown Market, Ranshet Wadi, Ra...  700025  22.530148   
25    Keoratala, Kalighat, Kalimandir, Protisthan, R...  700026  22.518749   
26    Alipore Bodyguard Line, Alipore Civil Court, A...  700027  22.525961   
27    Dum Dum, Jungipara Salgs Nager Bzr, Jugipara S...  700028  22.631986   
28    Dover Lane, Lake Market, Rash Behari Avenue, S...  700029  22.517586   
29                Ghughudanga, Purbe Sinthee, Sethbagan  700030  22.620038   
...                                                 ...     ...        ...   
1355                                              Metia  743440  -1.000000   
1356                                           Minakhan  743441  -1.000000   
1357                                       Nazathtkhola  743442  22.448604   
1358                                            Hadipur  743443  -1.000000   
1359                                 Achipur, Bansdroni  743501  -1.000000   
1360                                            Bhangar  743502  22.494411   
1361                                 Bishnupur Baruipur  743503  22.360813   
1362                                              Falta  743504  22.299429   
1363                                          Jairampur  743507  22.444075   
1364                                         I I M Joka  743512  -1.000000   
1365                                           Fathepur  743513  22.312735   
1366                                           Chaberia  743514  -1.000000   
1367                                              Bodra  743517  -1.000000   
1368                                           Gokarnee  743601  22.269304   
1369                                         Rangabelia  743602  -1.000000   
1370                                            Kautala  743603  22.045793   
1371                                          Gilarchat  743604  -1.000000   
1372                                              Nalua  743605  -1.000000   
1373                                         Gangasagar  743606  21.649135   
1374                                          Jalaberia  743607  -1.0

## Remove areas with undefined latitude/longitude

In [118]:
#remove areas with undefined latitude/longitude
kolkata_df = kolkata_df[kolkata_df.Latitude != -1]
kolkata_df

AreaName PinCode   Latitude  \
0     Ananda Bazar Patrika, Braborne Road, Council H...  700001  22.573613   
1     Cossipore Gun Factory, Cossipore H O, Indian R...  700002  22.618940   
2      Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  700003  22.603822   
3     R G Kar M C Hospital, Shyambazar, Shyambazar M...  700004  22.598321   
4                                   Ahiritola, Hatkhola  700005  22.598068   
5     Beadon Street, Bidhan Sarani, Jorasanko, Manik...  700006  22.589073   
6     Archana, Barabazar H O, College Squire, Kalake...  700007  22.582557   
7                                Barisha, Purba Barisha  700008  22.481168   
8                                     Sarani Rarsibagan  700009  22.576675   
9             Beleghata H O, Joranandir, Subhas Sarobar  700010  22.563025   
10                                          Narkeldanga  700011  22.577486   
11                               Bow Bazar, Sankaritala  700012  22.568353   
12                                Dharamtala, Mott Lane  700013  22.565300   
13    Bamboovilla, Intally, Linton Street, Sealdah, ...  700014  22.559624   
14                Seal Lane, Sales Tax Building, Tengra  700015  22.555040   
15               Elliot Road, Madrassa, Park Street H O  700016  22.553505   
16    Circus Avenue, Jhowtala, Park Circus, Shakespe...  700017  22.540940   
17                               Bartala, Rabindranagar  700018  22.536034   
18    Ballygunge, Ballygunge R S, Ballygunge Science...  700019  22.527999   
19    Archarya Jagadish, Chandra Bose Road, Gokhale ...  700020  22.537763   
20                                         Fort Willium  700021  22.563608   
21                                Bakery Road, Hastings  700022  22.548246   
22    Kidderpore, Manstala, Market, Raja Janakinath,...  700023  22.531739   
23    Garden Reach, Keshoram Cotton Mills, Paharpur ...  700024  22.548480   
24    Bhawanipore, Lansdown Market, Ranshet Wadi, Ra...  700025  22.530148   
25    Keoratala, Kalighat, Kalimandir, Protisthan, R...  700026  22.518749   
26    Alipore Bodyguard Line, Alipore Civil Court, A...  700027  22.525961   
27    Dum Dum, Jungipara Salgs Nager Bzr, Jugipara S...  700028  22.631986   
28    Dover Lane, Lake Market, Rash Behari Avenue, S...  700029  22.517586   
29                Ghughudanga, Purbe Sinthee, Sethbagan  700030  22.620038   
...                                                 ...     ...        ...   
1341                                       Basirhat R S  743414  22.655834   
1342                                              Hebia  743422  22.609947   
1343                                            Deganga  743423  22.685414   
1344                                           Devalaya  743424  22.733759   
1345                                              Haria  743425  22.543059   
1346                                           Hansabad  743426  22.532777   
1347                                           Katiahat  743427  22.734901   
1348                                          Kholapota  743428  22.675111   
1349                                  Taki Market, Taki  743429  22.604284   
1350                                         Hinglaganj  743435  22.449660   
1351                                             Bhabla  743436  22.554386   
1352                                        Dhanyakuria  743437  22.692588   
1353                                             Kolsur  743438  22.793834   
1354                                          Hatgachha  743439  22.247199   
1357                                       Nazathtkhola  743442  22.448604   
1360                                            Bhangar  743502  22.494411   
1361                                 Bishnupur Baruipur  743503  22.360813   
1362                                              Falta  743504  22.299429   
1363                                          Jairampur  743507  22.444075   
1365                                           Fathepur  743513  22.3

## Save processed data upto this point for future use

In [119]:
kolkata_df.to_csv('wb_area_pin_lat_lng.csv')

## Map the different pincodes on West Bengal map

In [19]:
import folium

In [24]:
kol_latitude = 22.5726; kol_longitude = 88.3639 #From Google search
map_kolkata = folium.Map(location = [kol_latitude, kol_longitude], zoom_start = 7)

# add markers to map
for lat, lng, pincode, area in zip(kolkata_df['Latitude'], kolkata_df['Longitude'], kolkata_df['PinCode'], kolkata_df['AreaName']):
    label = '{}, {}'.format(pincode, area)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_kolkata)  
    

map_kolkata

## Read the previously processed data into Dataframe for further processing

In [7]:
kolkata_df = pd.read_csv('wb_area_pin_lat_lng.csv', delimiter = ',')
kolkata_df = kolkata_df[[ 'PinCode', 'AreaName', 'Latitude', 'Longitude']]
kolkata_df                         

PinCode                                           AreaName   Latitude  \
0      700001  Ananda Bazar Patrika, Braborne Road, Council H...  22.573613   
1      700002  Cossipore Gun Factory, Cossipore H O, Indian R...  22.618940   
2      700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
3      700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
4      700005                                Ahiritola, Hatkhola  22.598068   
5      700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073   
6      700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
7      700008                             Barisha, Purba Barisha  22.481168   
8      700009                                  Sarani Rarsibagan  22.576675   
9      700010          Beleghata H O, Joranandir, Subhas Sarobar  22.563026   
10     700011                                        Narkeldanga  22.577486   
11     700012                             Bow Bazar, Sankaritala  22.568353   
12     700013                              Dharamtala, Mott Lane  22.565300   
13     700014  Bamboovilla, Intally, Linton Street, Sealdah, ...  22.559624   
14     700015              Seal Lane, Sales Tax Building, Tengra  22.555040   
15     700016             Elliot Road, Madrassa, Park Street H O  22.553505   
16     700017  Circus Avenue, Jhowtala, Park Circus, Shakespe...  22.540940   
17     700018                             Bartala, Rabindranagar  22.536034   
18     700019  Ballygunge, Ballygunge R S, Ballygunge Science...  22.527999   
19     700020  Archarya Jagadish, Chandra Bose Road, Gokhale ...  22.537763   
20     700021                                       Fort Willium  22.563608   
21     700022                              Bakery Road, Hastings  22.548246   
22     700023  Kidderpore, Manstala, Market, Raja Janakinath,...  22.531739   
23     700024  Garden Reach, Keshoram Cotton Mills, Paharpur ...  22.548480   
24     700025  Bhawanipore, Lansdown Market, Ranshet Wadi, Ra...  22.530148   
25     700026  Keoratala, Kalighat, Kalimandir, Protisthan, R...  22.518749   
26     700027  Alipore Bodyguard Line, Alipore Civil Court, A...  22.525961   
27     700028  Dum Dum, Jungipara Salgs Nager Bzr, Jugipara S...  22.631986   
28     700029  Dover Lane, Lake Market, Rash Behari Avenue, S...  22.517586   
29     700030              Ghughudanga, Purbe Sinthee, Sethbagan  22.620038   
...       ...                                                ...        ...   
1128   743414                                       Basirhat R S  22.655834   
1129   743422                                              Hebia  22.609947   
1130   743423                                            Deganga  22.685414   
1131   743424                                           Devalaya  22.733759   
1132   743425                                              Haria  22.543059   
1133   743426                                           Hansabad  22.532777   
1134   743427                                           Katiahat  22.734902   
1135   743428                                          Kholapota  22.675111   
1136   743429                                  Taki Market, Taki  22.604284   
1137   743435                                         Hinglaganj  22.449660   
1138   743436                                             Bhabla  22.554386   
1139   743437                                        Dhanyakuria  22.692588   
1140   743438                                             Kolsur  22.793834   
1141   743439                                          Hatgachha  22.247199   
1142   743442                                       Nazathtkhola  22.448605   
1143   743502                                            Bhangar  22.494411   
1144   743503                                 Bishnupur Baruipur  22.360813   
1145   743504                                              Falta  22.299429   
1146   743507                                          J

## Calculate distance from Kolkata for all the areas

In [13]:
import geopy.distance

kol_latitude = 22.5726; kol_longitude = 88.3639 #From Google search
Area_Name = []
Pin_Code = []
Lat = []
Lng = []
Distance = []
for index, row in kolkata_df.iterrows():
    PinCode_temp = row['PinCode']
    AreaName_temp = row['AreaName']
    AreaName_temp = AreaName_temp.strip()
    Lat_temp = row['Latitude']
    Lng_temp = row['Longitude']
    distance = geopy.distance.geodesic((kol_latitude,kol_longitude), (Lat_temp,Lng_temp)).km
    Area_Name.append(AreaName_temp)
    Pin_Code.append(PinCode_temp)
    Lat.append(Lat_temp)
    Lng.append(Lng_temp)
    Distance.append(distance)
    
kolkata_df = pd.DataFrame({"PinCode":Pin_Code,"AreaName":Area_Name,"Latitude":Lat,"Longitude":Lng,"Distance":Distance})
kolkata_df

PinCode                                           AreaName   Latitude  \
0      700001  Ananda Bazar Patrika, Braborne Road, Council H...  22.573613   
1      700002  Cossipore Gun Factory, Cossipore H O, Indian R...  22.618940   
2      700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
3      700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
4      700005                                Ahiritola, Hatkhola  22.598068   
5      700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073   
6      700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
7      700008                             Barisha, Purba Barisha  22.481168   
8      700009                                  Sarani Rarsibagan  22.576675   
9      700010          Beleghata H O, Joranandir, Subhas Sarobar  22.563026   
10     700011                                        Narkeldanga  22.577486   
11     700012                             Bow Bazar, Sankaritala  22.568353   
12     700013                              Dharamtala, Mott Lane  22.565300   
13     700014  Bamboovilla, Intally, Linton Street, Sealdah, ...  22.559624   
14     700015              Seal Lane, Sales Tax Building, Tengra  22.555040   
15     700016             Elliot Road, Madrassa, Park Street H O  22.553505   
16     700017  Circus Avenue, Jhowtala, Park Circus, Shakespe...  22.540940   
17     700018                             Bartala, Rabindranagar  22.536034   
18     700019  Ballygunge, Ballygunge R S, Ballygunge Science...  22.527999   
19     700020  Archarya Jagadish, Chandra Bose Road, Gokhale ...  22.537763   
20     700021                                       Fort Willium  22.563608   
21     700022                              Bakery Road, Hastings  22.548246   
22     700023  Kidderpore, Manstala, Market, Raja Janakinath,...  22.531739   
23     700024  Garden Reach, Keshoram Cotton Mills, Paharpur ...  22.548480   
24     700025  Bhawanipore, Lansdown Market, Ranshet Wadi, Ra...  22.530148   
25     700026  Keoratala, Kalighat, Kalimandir, Protisthan, R...  22.518749   
26     700027  Alipore Bodyguard Line, Alipore Civil Court, A...  22.525961   
27     700028  Dum Dum, Jungipara Salgs Nager Bzr, Jugipara S...  22.631986   
28     700029  Dover Lane, Lake Market, Rash Behari Avenue, S...  22.517586   
29     700030              Ghughudanga, Purbe Sinthee, Sethbagan  22.620038   
...       ...                                                ...        ...   
1128   743414                                       Basirhat R S  22.655834   
1129   743422                                              Hebia  22.609947   
1130   743423                                            Deganga  22.685414   
1131   743424                                           Devalaya  22.733759   
1132   743425                                              Haria  22.543059   
1133   743426                                           Hansabad  22.532777   
1134   743427                                           Katiahat  22.734902   
1135   743428                                          Kholapota  22.675111   
1136   743429                                  Taki Market, Taki  22.604284   
1137   743435                                         Hinglaganj  22.449660   
1138   743436                                             Bhabla  22.554386   
1139   743437                                        Dhanyakuria  22.692588   
1140   743438                                             Kolsur  22.793834   
1141   743439                                          Hatgachha  22.247199   
1142   743442                                       Nazathtkhola  22.448605   
1143   743502                                            Bhangar  22.494411   
1144   743503                                 Bishnupur Baruipur  22.360813   
1145   743504                                              Falta  22.299429   
1146   743507                                          J

In [14]:
kolkata_df.to_csv('wb_area_pin_lat_lng_dist.csv')

## Filter area which are within 20 km from Kolkata

In [17]:
kolkata_20_df = kolkata_df[kolkata_df.Distance <= 20]
kolkata_20_df = kolkata_20_df.reset_index()
kolkata_20_df

index  PinCode                                           AreaName  \
0        0   700001  Ananda Bazar Patrika, Braborne Road, Council H...   
1        1   700002  Cossipore Gun Factory, Cossipore H O, Indian R...   
2        2   700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika   
3        3   700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...   
4        4   700005                                Ahiritola, Hatkhola   
5        5   700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...   
6        6   700007  Archana, Barabazar H O, College Squire, Kalake...   
7        7   700008                             Barisha, Purba Barisha   
8        8   700009                                  Sarani Rarsibagan   
9        9   700010          Beleghata H O, Joranandir, Subhas Sarobar   
10      10   700011                                        Narkeldanga   
11      11   700012                             Bow Bazar, Sankaritala   
12      12   700013                              Dharamtala, Mott Lane   
13      13   700014  Bamboovilla, Intally, Linton Street, Sealdah, ...   
14      14   700015              Seal Lane, Sales Tax Building, Tengra   
15      15   700016             Elliot Road, Madrassa, Park Street H O   
16      16   700017  Circus Avenue, Jhowtala, Park Circus, Shakespe...   
17      17   700018                             Bartala, Rabindranagar   
18      18   700019  Ballygunge, Ballygunge R S, Ballygunge Science...   
19      19   700020  Archarya Jagadish, Chandra Bose Road, Gokhale ...   
20      20   700021                                       Fort Willium   
21      21   700022                              Bakery Road, Hastings   
22      22   700023  Kidderpore, Manstala, Market, Raja Janakinath,...   
23      23   700024  Garden Reach, Keshoram Cotton Mills, Paharpur ...   
24      24   700025  Bhawanipore, Lansdown Market, Ranshet Wadi, Ra...   
25      25   700026  Keoratala, Kalighat, Kalimandir, Protisthan, R...   
26      26   700027  Alipore Bodyguard Line, Alipore Civil Court, A...   
27      27   700028  Dum Dum, Jungipara Salgs Nager Bzr, Jugipara S...   
28      28   700029  Dover Lane, Lake Market, Rash Behari Avenue, S...   
29      29   700030              Ghughudanga, Purbe Sinthee, Sethbagan   
..     ...      ...                                                ...   
157    168   711224                                            Dankuni   
158    171   711227                                          Ghoshpara   
159    173   711302                                        Andul Mauri   
160    180   711309                                        Delta Mills   
161    182   711311                               Jagaeha, Old Jagacha   
162    184   711313                                           Sankrail   
163    187   711321                      G I Press Colony, Old Jagacha   
164    192   711402                                          Baltikuri   
165    193   711403                                               Bank   
166    195   711405                                             Domjur   
167    197   711409                                           Makardah   
168    199   711411                                              Begri   
169    226   712201  Kalbazar, Manicktala Bazar, Serampore Court, S...   
170    227   712202                                            Mahes-I   
171    228   712203                                        Mallickpara   
172    233   712232                                         Bhadrakali   
173    234   712233                                         Hind Motor   
174    235   712234                                           Kanaipur   
175    236   712235                              Cripar Road, Konnagar   
176    237   712245                                             Makhla   
177    238   712246                                 Nabagram (Hooghly)   
178    239   712247                             Raghunathpur (Hooghly)   
179    24

## Map to show areas within 20 KM from centre of Kolkata

In [25]:
kol_latitude = 22.5726; kol_longitude = 88.3639 #From Google search
map_kolkata = folium.Map(location = [kol_latitude, kol_longitude], zoom_start = 10.7)

# add markers to map
for lat, lng, pincode, area in zip(kolkata_20_df['Latitude'], kolkata_20_df['Longitude'], kolkata_20_df['PinCode'], kolkata_20_df['AreaName']):
    label = '{}, {}'.format(pincode, area)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_kolkata)  
    

map_kolkata

## Populate relevant venues nearby all the pincodes selected

In [49]:
CLIENT_ID = 'J1JBTE1UXQVN0P0W2TV04LA2HMJ3ZYMOY0QOA42IWYCIG0CW'
CLIENT_SECRET = 'XWM0RWPJN11YUG34VHB1QQI0FSPBJOYE3AJ2RAHCQTIWRDUZ' # Removed manually after use
VERSION = '20190308' 
LIMIT = 100
radius = 500

In [50]:
import json # library to handle JSON files
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [52]:
Pin_Code = []
Area_Name = []
Lat = []
Lng = []
Distance = []
Venue_name = []
Venue_cat = []
Venue_lat = []
Venue_lng = []

for index, row in kolkata_20_df.iterrows():
    PinCode_temp = row['PinCode']
    AreaName_temp = row['AreaName']
    Lat_temp = row['Latitude']
    Lng_temp = row['Longitude']
    dist_temp = row['Distance']
    
    try:
        url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, Lat_temp, Lng_temp, VERSION, radius, LIMIT)
        results = requests.get(url).json()
    
        venues = results['response']['groups'][0]['items']  
        nearby_venues = json_normalize(venues) # flatten JSON
    
        # filter the category for each row
        nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

        # clean columns
        nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]
        for index1, row1 in nearby_venues.iterrows():
            Pin_Code.append(PinCode_temp)
            Area_Name.append(AreaName_temp)
            Lat.append(Lat_temp)
            Lng.append(Lng_temp)
            Distance.append(dist_temp)
            name_temp = row1['name']
            categories_temp = row1['categories']
            lat1_temp = row1['lat']
            lng1_temp = row1['lng']
            Venue_name.append(name_temp)
            Venue_cat.append(categories_temp)
            Venue_lat.append(lat1_temp)
            Venue_lng.append(lng1_temp)
    except:
        Pin_Code.append(PinCode_temp)
        Area_Name.append(AreaName_temp)
        Lat.append(Lat_temp)
        Lng.append(Lng_temp)
        Distance.append(dist_temp)
        name_temp = row1['name']
        Venue_name.append('None')
        Venue_cat.append('None')
        Venue_lat.append(-1)
        Venue_lng.append(-1)
        
        
kolkata_20_df = pd.DataFrame({"PinCode":Pin_Code,"AreaName":Area_Name,"Latitude":Lat,"Longitude":Lng,"Distance":Distance,"VenueName":Venue_name,"Category":Venue_cat,"VenueLatitude":Venue_lat,"VenueLongitude":Venue_lng})
kolkata_20_df        
        
        

PinCode                                           AreaName   Latitude  \
0     700001  Ananda Bazar Patrika, Braborne Road, Council H...  22.573613   
1     700001  Ananda Bazar Patrika, Braborne Road, Council H...  22.573613   
2     700001  Ananda Bazar Patrika, Braborne Road, Council H...  22.573613   
3     700002  Cossipore Gun Factory, Cossipore H O, Indian R...  22.618940   
4     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
5     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
6     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
7     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
8     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
9     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
10    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
11    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
12    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
13    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
14    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
15    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
16    700005                                Ahiritola, Hatkhola  22.598068   
17    700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073   
18    700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073   
19    700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073   
20    700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073   
21    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
22    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
23    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
24    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
25    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
26    700008                             Barisha, Purba Barisha  22.481168   
27    700008                             Barisha, Purba Barisha  22.481168   
28    700008                             Barisha, Purba Barisha  22.481168   
29    700009                                  Sarani Rarsibagan  22.576675   
..       ...                                                ...        ...   
566   712201  Kalbazar, Manicktala Bazar, Serampore Court, S...  22.748431   
567   712201  Kalbazar, Manicktala Bazar, Serampore Court, S...  22.748431   
568   712202                                            Mahes-I  22.735090   
569   712203                                        Mallickpara  22.745277   
570   712232                                         Bhadrakali  22.673991   
571   712233                                         Hind Motor  22.685856   
572   712234                                           Kanaipur  22.697519   
573   712235                              Cripar Road, Konnagar  22.700774   
574   712245                                             Makhla  22.672289   
575   712246                                 Nabagram (Hooghly)  22.697671   
576   712247                             Raghunathpur (Hooghly)  22.657546   
577   712248  Bangurpark, Gandhisarak, Mahesh-Ii, Masjidlane...  22.723061   
578   712258        Raiendra Avenue, Uttarpara, Uttarpara Bazar  22.666507   
579   712258        Raiendra Avenue, Uttarpara, Uttarpara Bazar  22.666507   
580   712304                                              Janai  22.717670   
581   712702                                         Chanditala  22.696795   
582   712703                                           Nawabpur  22.696795   
583   743178  Ghola, Housing Estate, Ramchandrapur, Tirthaba...  22.697172   
584   743178  Ghola, Housing Estate, Ramchandrapur, Tirthaba...  22.697172   
585   743178  Ghola, Housin

In [53]:
kolkata_20_df.to_csv('wb_area_pin_lat_lng_20km_venues.csv')

## Read the previously processed data into Dataframe for further processing

In [3]:
kolkata_20_df = pd.read_csv('wb_area_pin_lat_lng_20km_venues.csv', delimiter = ',')
kolkata_20_df = kolkata_20_df[[ 'PinCode', 'AreaName', 'Latitude', 'Longitude','Distance','VenueName','Category','VenueLatitude','VenueLongitude']]
kolkata_20_df 

PinCode                                           AreaName   Latitude  \
0     700001  Ananda Bazar Patrika, Braborne Road, Council H...  22.573613   
1     700001  Ananda Bazar Patrika, Braborne Road, Council H...  22.573613   
2     700001  Ananda Bazar Patrika, Braborne Road, Council H...  22.573613   
3     700002  Cossipore Gun Factory, Cossipore H O, Indian R...  22.618940   
4     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
5     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
6     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
7     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
8     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
9     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
10    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
11    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
12    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
13    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
14    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
15    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
16    700005                                Ahiritola, Hatkhola  22.598068   
17    700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073   
18    700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073   
19    700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073   
20    700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073   
21    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
22    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
23    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
24    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
25    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
26    700008                             Barisha, Purba Barisha  22.481168   
27    700008                             Barisha, Purba Barisha  22.481168   
28    700008                             Barisha, Purba Barisha  22.481168   
29    700009                                  Sarani Rarsibagan  22.576675   
..       ...                                                ...        ...   
566   712201  Kalbazar, Manicktala Bazar, Serampore Court, S...  22.748431   
567   712201  Kalbazar, Manicktala Bazar, Serampore Court, S...  22.748431   
568   712202                                            Mahes-I  22.735090   
569   712203                                        Mallickpara  22.745277   
570   712232                                         Bhadrakali  22.673991   
571   712233                                         Hind Motor  22.685856   
572   712234                                           Kanaipur  22.697519   
573   712235                              Cripar Road, Konnagar  22.700774   
574   712245                                             Makhla  22.672289   
575   712246                                 Nabagram (Hooghly)  22.697671   
576   712247                             Raghunathpur (Hooghly)  22.657546   
577   712248  Bangurpark, Gandhisarak, Mahesh-Ii, Masjidlane...  22.723061   
578   712258        Raiendra Avenue, Uttarpara, Uttarpara Bazar  22.666507   
579   712258        Raiendra Avenue, Uttarpara, Uttarpara Bazar  22.666507   
580   712304                                              Janai  22.717671   
581   712702                                         Chanditala  22.696795   
582   712703                                           Nawabpur  22.696795   
583   743178  Ghola, Housing Estate, Ramchandrapur, Tirthaba...  22.697172   
584   743178  Ghola, Housing Estate, Ramchandrapur, Tirthaba...  22.697172   
585   743178  Ghola, Housin

## Places do have Parks for children

In [24]:
kolkata_20_park_df = kolkata_20_df[kolkata_20_df.Category == "Park"]
kolkata_20_park_df

PinCode                                           AreaName   Latitude  \
0     700001  Ananda Bazar Patrika, Braborne Road, Council H...  22.573613   
5     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
16    700005                                Ahiritola, Hatkhola  22.598068   
23    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
25    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
36    700012                             Bow Bazar, Sankaritala  22.568353   
52    700013                              Dharamtala, Mott Lane  22.565300   
59    700015              Seal Lane, Sales Tax Building, Tengra  22.555040   
92    700016             Elliot Road, Madrassa, Park Street H O  22.553505   
108   700017  Circus Avenue, Jhowtala, Park Circus, Shakespe...  22.540940   
144   700020  Archarya Jagadish, Chandra Bose Road, Gokhale ...  22.537763   
151   700021                                       Fort Willium  22.563608   
170   700029  Dover Lane, Lake Market, Rash Behari Avenue, S...  22.517586   
291   700064                                       Bidhan Nagar  22.592457   
293   700064                                       Bidhan Nagar  22.592457   
355   700072  Hindustan Building, Princeep Street, Sooterkee...  22.565300   
365   700073  Colootola, Chittaranjan Avenue, Kolkata Univer...  22.574311   
431   700091                                        Sech Bhavan  22.581696   

     Longitude  Distance                          VenueName Category  \
0    88.348341  1.604059                    Millennium Park     Park   
5    88.367960  3.482579       Bagh Bazar Durga Puja Ground     Park   
16   88.358119  2.882242                     Kumartuli Park     Park   
23   88.361703  1.125577                        Girish Park     Park   
25   88.361703  1.125577          Victoria Memorial Grounds     Park   
36   88.361571  0.527732               Santosh Mitra Square     Park   
52   88.354289  1.276920                  Wellington Square     Park   
59   88.381247  2.639025                     Seal Lane Maat     Park   
92   88.356574  2.244778  Allen Park (Gopinath Sahay Udyan)     Park   
108  88.368135  3.532956                 Park Circus Maidan     Park   
144  88.351468  4.064120      Subhash Udyan (Northern Park)     Park   
151  88.338588  2.787156                   Eden Garden Park     Park   
170  88.349433  6.271225                    Deshapriya Park     Park   
291  88.412665  5.475682                      BF Block Park     Park   
293  88.412665  5.475682                      BH Block Park     Park   
355  88.354289  1.276920                  Wellington Square     Park   
365  88.361571  0.305403                       Md. Ali Park     Park   
431  88.430414  6.914019                          CKCL Park     Park   

     VenueLatitude  VenueLongitude  
0        22.572481       88.344517  
5        22.604344       88.366277  
16       22.598956       88.361514  
23       22.586069       88.362404  
25       22.585984       88.363124  
36       22.566008       88.365719  
52       22.563509       88.357525  
59       22.551997       88.382552  
92       22.550274       88.354282  
108      22.542981       88.368702  
144      22.535204       88.350350  
151      22.564244       88.340743  
170      22.518395       88.353177  
291      22.594503       88.414827  
293      22.594553       88.414876  
355      22.563509       88.357525  
365      22.577500       88.360720  
431      22.584534       88.427582

## Finally the places where need to build new Parks for children

In [25]:
kolkata_20_park_pins = kolkata_20_park_df['PinCode'].tolist()
kolkata_20_park_pins
kolkata_20_new_park_df = kolkata_20_df[["PinCode","AreaName"]].drop_duplicates()
kolkata_20_new_park_df = kolkata_20_new_park_df[~kolkata_20_new_park_df["PinCode"].isin(kolkata_20_park_pins)]
kolkata_20_new_park_df

PinCode                                           AreaName
3     700002  Cossipore Gun Factory, Cossipore H O, Indian R...
10    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...
17    700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...
26    700008                             Barisha, Purba Barisha
29    700009                                  Sarani Rarsibagan
31    700010          Beleghata H O, Joranandir, Subhas Sarobar
34    700011                                        Narkeldanga
56    700014  Bamboovilla, Intally, Linton Street, Sealdah, ...
114   700018                             Bartala, Rabindranagar
115   700019  Ballygunge, Ballygunge R S, Ballygunge Science...
152   700022                              Bakery Road, Hastings
153   700023  Kidderpore, Manstala, Market, Raja Janakinath,...
154   700024  Garden Reach, Keshoram Cotton Mills, Paharpur ...
155   700025  Bhawanipore, Lansdown Market, Ranshet Wadi, Ra...
157   700026  Keoratala, Kalighat, Kalimandir, Protisthan, R...
161   700027  Alipore Bodyguard Line, Alipore Civil Court, A...
166   700028  Dum Dum, Jungipara Salgs Nager Bzr, Jugipara S...
181   700030              Ghughudanga, Purbe Sinthee, Sethbagan
185   700031                             Dhakuria, K P Roy Lane
187   700032  Bejoy Garh, Hossain Road, Jadavpur University,...
191   700033  Indrani Park, Tallygunge H O, Tollygunge Circu...
196   700034  Brahma Samaj Road, Behala, Behala Municipal Ma...
200   700035  Alambazar, Ashokegarh, Bon Hooghly, Desh Bandh...
204   700036                         Bengal Immunity, Baranagar
205   700037         Belgachia, Belgachia Road, Northern Avenue
207   700038                                            Sahapur
210   700039  Bediadanga, Picnic Garden, Topsia, Tiljala, Ti...
217   700040                    Netajinagar, Regent Park, Russa
221   700041                                    Paschim Putuary
223   700042                             Bose Pukur Road, Kasba
..       ...                                                ...
553   711224                                            Dankuni
554   711227                                          Ghoshpara
555   711302                                        Andul Mauri
557   711309                                        Delta Mills
558   711311                               Jagaeha, Old Jagacha
559   711313                                           Sankrail
560   711321                      G I Press Colony, Old Jagacha
561   711402                                          Baltikuri
562   711403                                               Bank
563   711405                                             Domjur
564   711409                                           Makardah
565   711411                                              Begri
566   712201  Kalbazar, Manicktala Bazar, Serampore Court, S...
568   712202                                            Mahes-I
569   712203                                        Mallickpara
570   712232                                         Bhadrakali
571   712233                                         Hind Motor
572   712234                                           Kanaipur
573   712235                              Cripar Road, Konnagar
574   712245                                             Makhla
575   712246                                 Nabagram (Hooghly)
576   712247                             Raghunathpur (Hooghly)
577   712248  Bangurpark, Gandhisarak, Mahesh-Ii, Masjidlane...
578   712258        Raiendra Avenue, Uttarpara, Uttarpara Bazar
580   712304                                              Janai
581   712702                                         Chanditala
582   712703                                           Nawabpur
583   743178  Ghola, Housing Estate, Ramchandrapur, Tirthaba...
586   743201                           Barasat Bazar, Sethpukur
591   743319                                    B B Natun Bazar

[171 rows x 2 columns]

# So there are 171 new Children Parks will be recommended to the government. The above location information will be handed over to the Kolkata Corporation.